tests/clickhouse/test_formatter.py

```diff
@@ -0,0 +1,140 @@
+    ),  # Hierarchical function call
+    (
+        FunctionCall(
+            None,
+            "f1",
+            [
+                FunctionCall("al1", "f2", [Column(None, "param1", "table1")]),
+                FunctionCall("al2", "f3", [Column(None, "param2", "table1")]),
+            ],
+        ),
+        "f1((f2(table1.param1) AS al1), (f3(table1.param2) AS al2))",
+    ),  # Hierarchical function call with aliases
+    (
+        CurriedFunctionCall(
+            None,
+            FunctionCall(None, "f0", [Column(None, "param1", "table1")]),
+            [
+                FunctionCall(None, "f1", [Column(None, "param2", "table1")]),
+                Column(None, "param3", "table1"),
+            ],
+        ),
+        "f0(table1.param1)(f1(table1.param2), table1.param3)",
+    ),  # Curried function call with hierarchy
+]
+
+
+@pytest.mark.parametrize("expression, expected", test_expressions)
+def test_format_expressions(expression: Expression, expected: str) -> None:
+    visitor = ClickhouseExpressionFormatter()
+    assert expression.accept(visitor) == expected
+
+
+def test_aliases() -> None:
+    # No context
+    col1 = Column("al1", "column1", "table1")
+    col2 = Column("al1", "column1", "table1")
+
+    assert col1.accept(ClickhouseExpressionFormatter()) == "(table1.column1 AS al1)"
+    assert col2.accept(ClickhouseExpressionFormatter()) == "(table1.column1 AS al1)"
+
+    # With Context
+    pc = ParsingContext()
+    assert col1.accept(ClickhouseExpressionFormatter(pc)) == "(table1.column1 AS al1)"
+    assert col2.accept(ClickhouseExpressionFormatter(pc)) == "al1"
+
+    # Hierarchical expression inherits parsing context and applies alaises
+    f = FunctionCall(
+        None,
+        "f1",
+        [
+            FunctionCall("tag[something]", "tag", [Column(None, "column1", "table1")]),
+            FunctionCall("tag[something]", "tag", [Column(None, "column1", "table1")]),
+            FunctionCall("tag[something]", "tag", [Column(None, "column1", "table1")]),
+        ],
+    )
+
+    expected = "f1((tag(table1.column1) AS `tag[something]`), `tag[something]`, `tag[something]`)"
+    assert f.accept(ClickhouseExpressionFormatter()) == expected
+
+
+test_escaped = [
+    (
+        Column(None, "tags.values", "table.something"),
+        "table.something.tags.values",
+    ),  # Columns with dot are not escaped
+    (
+        Column(None, "tags[something]", "weird_!@#$%^^&*_table"),
+        "`weird_!@#$%^^&*_table`.`tags[something]`",
+    ),  # Somebody thought that table name was a good idea.
+    (
+        Column("alias.cannot.have.dot", "columns.can", "table"),
+        "(table.columns.can AS `alias.cannot.have.dot`)",
+    ),  # Escaping is different between columns and aliases
+    (
+        FunctionCall(None, "f*&^%$#unction", [Column(None, "column", "table")]),
+        "`f*&^%$#unction`(table.column)",
+    ),  # Function names can be escaped. Hopefully it will never happen
```

Comment:
I didn't realize this was actually valid syntax, but looks like it is. Interesting.

Reply:
I hope there is no clickhouse function that actually requires escaping. It would be ridiculous. So in practice this is unlikely to ever happen